chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py

```diff
@@ -4,98 +4,197 @@
-                file_status = os.stat(path)
 
-                if path in self.non_existent_paths:
-                    self.non_existent_paths.discard(path)
-                    daemon_log.debug('New device started to respond %s' % path)
+def find_device_by_device_path(p, xs):
+    return next(iter([d for d in xs if d['device_path'] == p]), None)
 
-                self.previous_path_status[path] = file_status
-                break
-            except OSError as os_error:
-                if os_error.errno not in [errno.ENOENT, errno.ENOTDIR]:
-                    raise
 
-                # An OSError could be raised because a path genuinely doesn't
-                # exist, but it also can be the result of conflicting with
-                # actions that cause devices to disappear momentarily, such as
-                # during a partprobe while it reloads the partition table.
-                # So we retry for a short window to catch those devices that
-                # just disappear momentarily.
-                time.sleep(0.1)
-                retries -= retries if path in self.non_existent_paths else 1
+def mutate_parent_prop(xs):
+    disks = [x for x in xs if x['device_type'] == 'disk']
+    partitions = [x for x in xs if x['device_type'] == 'partition']
 
-        if file_status is None:
-            if path not in self.non_existent_paths:
-                self.non_existent_paths.add(path)
-                daemon_log.debug('New device failed to respond %s' % path)
+    for x in partitions:
+        parent_path = get_parent_path(x['device_path'])
+        device = find_device_by_device_path(parent_path, disks)
 
-            if path not in self.previous_path_status:
-                return None
+        if device:
+            x['parent'] = device['major_minor']
 
-            file_status = self.previous_path_status.pop(path)
-            daemon_log.debug('Device failed to respond but stored file_status used')
 
-        if stat.S_ISBLK(file_status.st_mode):
-            return "%d:%d" % (os.major(file_status.st_rdev), os.minor(file_status.st_rdev))
-        else:
-            return None
+def filter_device(x):
+    # Exclude zero-sized devices
+    if x['size'] == 0:
+        return False
+
+    # Exclude ramdisks, floppy drives, obvious cdroms
+    if re.search("^ram\d+$", x['device_path']) or\
+       re.search("^fd\d+$", x['device_path']) or\
+       re.search("^sr\d+$", x['device_path']):
+        return False
+
+    # Exclude read-only devices and removed media or devices
+    try:
+        # Never use 'w' in the built-in open() or it'll create
+        # a 0 length file where a
+        # device was removed!
+        fd = os.open("/dev/%s" % x['device_path'], os.O_WRONLY)
+    except OSError, e:
+        # EROFS: Device is read-only
+        # ENOENT: No such file or directory
+        # NO_MEDIA_ERRNO: No medium found
+        if e.errno in [errno.EROFS, errno.ENOENT, errno.NO_MEDIA_ERRNO]:
+            return False
```

Comment:
what about when e.errno not in that list, don't you want to raise? do you want to use the device

Reply:
I think I can replace all of this with some udev rules.